### PR TITLE
feature: skip & rerun failed tasks

### LINKED ISSUE #2499: [Feature][Config UI] Skipping Failed/Stalled Stages
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-devlake/issues?q=is%3Aissue) and found no similar feature requirement.

### Description

The requested feature is an option to skip failed stages and carry on with the following stages in the pipeline, for those who have a long list of repos to run and want to make as much progress as possible in one go.

In "advanced mode", one can list tens or hundreds of stages(repos) in the "2D list" of the API, to run in sequence. But when one stage fails for whatever reason, the whole pipeline halts, missing the opportunity to run the following stages and just wasting time for human intervention.

Better still, a timeout (say 4 hours) setting can also be useful, for those stages that don't actually fail but still are not making any progress due to rate limiting etc.

### Use case

Add a global option "skip stage when stalled", and a sub-setting for a timeout.

### Related issues

_No response_

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by mindlesscloud**
# Summary
Give users the option to skip/rerun failed tasks, and rerun them in the future.
To skip failed tasks, set the `onFailSkip=true` on creating the blueprint.
image
To rerun failed tasks, post `taskId` to the API `/pipelines/{pipelineId}/tasks`, if `taskId = 0` all failed tasks will rerun
image
To get tasks, `GET /pipelines/{pipelineId}/tasks`, only the most recent tasks will be returned

## The process flow for rerun:
**We introduced a new pipeline status `TASK_RERUN` which indicates the pipeline is going to rerun**
### The HTTP handler part
1. find the pipeline by ID, and return an error if the status is among `TASK_CREATED`, `TASK_RUNNING`, and `TASK_RERUN`
2. find the tasks that need to be rerun
3. delete all tasks with the status `TASK_CREATED`
4. generate new tasks from the tasks that need to be rerun, and set their status as `TASK_CREATED`
5. set the pipeline status as `TASK_RERUN`

### The pipeline scheduler part

1. scan the table `_devlake_pipelines`, and find the first pipeline with status `TASK_CREATED` or `TASK_RERUN`
2. set the status of the pipeline to `TASK_RUNNING`
3. run the pipeline in another goroutine

### Does this close any open issues?
Closes #2499

### CODE REVIEW

**Comment by mintsweet** [comment]
@mindlesscloud 
The design of the product also expects a rerun of a single task, regardless of whether the task failed or not.

screenshot-20221107-103159

File: api/task/task.go [hunk 1]
```diff
@@ -23,67 +23,120 @@ import (
...
 
-/*
-Get list of pipelines
-GET /pipelines/pipeline:id/tasks?status=TASK_RUNNING&pending=1&page=1&=pagesize=10
-{
-	"tasks": [
-		{"id": 1, "plugin": "", ...}
-	],
-	"count": 5
+func Delete(c *gin.Context) {
```
Comment (likyh) [inline api/task/task.go:31]: what's this for?

**Comment by likyh** [comment]
Can I use skipOnFail in advance mode?

File: services/task.go [hunk 2]
```diff
@@ -194,6 +196,52 @@ func GetTasks(query *TaskQuery) ([]models.Task, int64, errors.Error) {
...
+	return result, nil
+}
+
+// SpawnTasks create new tasks from the failed ones
+func SpawnTasks(input []models.Task) ([]models.Task, errors.Error) {
+	var result []models.Task
+	for _, task := range input {
+		task.Model = common.Model{}
+		task.Status = models.TASK_CREATED
+		result = append(result, task)
```
Comment (likyh) [inline services/task.go:235]: does these task need to show in blueprint detail page?

File: services/task.go [hunk 3]
```diff
@@ -194,6 +196,52 @@ func GetTasks(query *TaskQuery) ([]models.Task, int64, errors.Error) {
...
+	}
+	runningTasks.FillProgressDetailToTasks(result)
+	return result, nil
+}
+
+// SpawnTasks create new tasks from the failed ones
+func SpawnTasks(input []models.Task) ([]models.Task, errors.Error) {
+	var result []models.Task
+	for _, task := range input {
+		task.Model = common.Model{}
```
Comment (likyh) [inline services/task.go:233]: Just pick the necessary column here? It seems that some unnecessary columns like `FinishedAt`  have been copied to new task

File: api/task/task.go [hunk 4]
```diff
@@ -23,67 +23,120 @@ import (
...
+	newTasks, err := services.SpawnTasks(failedTasks)
+	if err != nil {
+		shared.ApiOutputError(c, errors.Default.Wrap(err, "error create tasks"))
+		return
+	}
+	for _, task := range newTasks {
+		taskIds = append(taskIds, task.ID)
+	}
+	go func() {
+		_ = services.RunTasksStandalone(services.GetPipelineLogger(pipeline), taskIds)
```
Comment (likyh) [inline api/task/task.go:139]: If run here direct, it will break running control by `PIPELINE_MAX_PARALLEL` and all tasks will run parallel.

**Review by likyh — CHANGES_REQUESTED** [review]

File: runner/run_pipeline.go [hunk 5]
```diff
@@ -37,43 +36,46 @@ func RunPipeline(
...
-		return errors.Internal.New(fmt.Sprintf("expected total tasks to be %v, got %v", dbPipeline.TotalTasks, len(tasks)))
-	}
-	// convert to 2d array
+	sort.Slice(tasks, func(i, j int) bool {
+		if tasks[i].PipelineRow < tasks[j].PipelineRow {
+			return true
+		} else if tasks[i].PipelineRow == tasks[j].PipelineRow {
+			return tasks[i].PipelineCol < tasks[j].PipelineCol
+		}
+		return true
```
Comment (likyh) [inline runner/run_pipeline.go:51]: `return false if tasks[i].PipelineRow < tasks[j].PipelineRow`?

File: services/task.go [hunk 6]
```diff
@@ -303,3 +361,12 @@ func getTaskIdFromActivityId(activityId string) (uint64, errors.Error) {
 	}
 	return errors.Convert01(strconv.ParseUint(submatches[1], 10, 64))
 }
+
+// DeleteCreatedTasks deletes tasks with status `TASK_RERUN`
```
Comment (likyh) [inline services/task.go:365]: DeleteCreatedTasks deletes `TASK_RERUN`?

File: services/pipeline.go [hunk 7]
```diff
@@ -168,24 +167,27 @@ func RunPipelineInQueue(pipelineMaxParallel int64) {
 		dbPipeline := &models.DbPipeline{}
 		for {
 			cronLocker.Lock()
-			db.Where("status = ?", models.TASK_CREATED).
-				Not(startedPipelineIds).
+			db.Where("status IN ?", []string{models.TASK_CREATED, models.TASK_RERUN}).
```
Comment (likyh) [inline services/pipeline.go:170]: Hmmm, there is some confusion about the status of the pipeline named TASK_XXX. But not written by this PR. Let's refactor this in another issue.

File: runner/run_pipeline.go [hunk 8]
```diff
@@ -90,18 +92,16 @@ func RunPipeline(
 			log.Error(err, "run tasks failed")
 			return errors.Convert(err)
 		}
-		// Deprecated
+
 		// update finishedTasks
-		finishedTasks += len(row)
 		err = db.Model(dbPipeline).Updates(map[string]interface{}{
-			"finished_tasks": finishedTasks,
+			"finished_tasks": gorm.Expr("finished_tasks + ?", len(row)),
```
Comment (likyh) [inline runner/run_pipeline.go:98]: How do the finished_tasks count when rerunning only one task?

File: runner/run_pipeline.go (hunk 5 above)
Comment (likyh) [reply runner/run_pipeline.go:51]: Let's add some unit-test here to confirm 2d array are generated as expected.

**Review by likyh — APPROVED** [review]
LGTM

**Pull request merged**